### PR TITLE
pivotal-165976451: Submitter security interface

### PR DESCRIPTION
**Pull request by Juan-EBI**
https://www.pivotaltracker.com/story/show/165976451
- added security login with end points

### CODE REVIEW

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/converters/AuxInfoConverter.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+import ac.uk.ebi.biostd.persistence.model.AuxInfo
+import java.io.StringReader
+import java.io.StringWriter
+import javax.persistence.AttributeConverter
+import javax.xml.bind.JAXBContext
+import javax.xml.bind.Marshaller
+
+class AuxInfoConverter : AttributeConverter<AuxInfo, String> {
+
```
Comment (jhoanmanuelms) [inline submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/converters/AuxInfoConverter.kt]: I thought we'd decided to not use this extra line.

File: submission/submission-security/src/main/kotlin/ebi/ac/uk/security/integration/SecurityEvents.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import ebi.ac.uk.security.integration.model.events.UserRegister
+import io.reactivex.Observable
+import io.reactivex.subjects.PublishSubject
+
+class SecurityEvents {
+
+    companion object {
+
+        // Subjects
+        internal val userPreRegister by lazy { PublishSubject.create<UserPreRegister>() }
```
Comment (jhoanmanuelms) [inline submission/submission-security/src/main/kotlin/ebi/ac/uk/security/integration/SecurityEvents.kt]: :+1:

File: submission/submission-security/src/main/kotlin/ebi/ac/uk/security/integration/SecurityEvents.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        internal val userRegister by lazy { PublishSubject.create<UserRegister>() }
+
+        // Observables
+        fun userPreRegister(): Observable<UserPreRegister> = userPreRegister
```
Comment (jhoanmanuelms) [inline submission/submission-security/src/main/kotlin/ebi/ac/uk/security/integration/SecurityEvents.kt]: What is pre-register? I'm assuming before activation? If so, probably we should use other name

File: submission/submission-security/src/main/kotlin/ebi/ac/uk/security/service/SecurityService.kt [hunk 4]
```diff
@@ -4,52 +4,73 @@ import ac.uk.ebi.biostd.persistence.model.User
...
+    private val userPreRegister: Subject<UserPreRegister> = SecurityEvents.userPreRegister,
+    private val userRegister: Subject<UserRegister> = SecurityEvents.userRegister
+) : ISecurityService {
+
+    override fun login(loginRequest: LoginRequest): UserInfo =
+        userRepository
+            .findByLoginOrEmailAndActive(loginRequest.login, loginRequest.login, true)
+            .filter { securityUtil.checkPassword(it.passwordDigest, loginRequest.password) }
+            .orElseThrow { LoginException() }
+            .let { UserInfo(securityUtil.createToken(it), it) }
```
Comment (jhoanmanuelms) [inline submission/submission-security/src/main/kotlin/ebi/ac/uk/security/service/SecurityService.kt]: Nice

File: submission/submission-security/src/main/kotlin/ebi/ac/uk/security/service/SecurityService.kt [hunk 5]
```diff
@@ -4,52 +4,73 @@ import ac.uk.ebi.biostd.persistence.model.User
...
+
+    override fun registerUser(request: RegisterRequest): User {
+        if (userRepository.existsByEmail(request.email)) {
```
Comment (jhoanmanuelms) [inline submission/submission-security/src/main/kotlin/ebi/ac/uk/security/service/SecurityService.kt]: Why don't you include this case also in the `when` clause?

File: submission/submission-security/src/test/kotlin/ebi/ac/uk/security/service/SecurityServiceTest.kt [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+) {
+
+    private val userPreRegister: Subject<UserPreRegister> = PublishSubject.create<UserPreRegister>()
+    private val userRegister: Subject<UserRegister> = PublishSubject.create<UserRegister>()
+
+    private val testInstance: SecurityService =
+        SecurityService(userRepository, securityUtil, securityProps, userPreRegister, userRegister)
+
+    @Test
+    fun login() {
```
Comment (jhoanmanuelms) [inline submission/submission-security/src/test/kotlin/ebi/ac/uk/security/service/SecurityServiceTest.kt]: We shouldn't have empty methods

File: submission/submission-security/src/test/kotlin/ebi/ac/uk/security/service/SecurityServiceTest.kt [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            every { securityProps.requireActivation } returns false
+            every { securityUtil.newKey() } returns SECRET_KEY
+
+            val subscriber = TestObserver<UserRegister>()
+            userRegister.subscribe(subscriber)
+
+            testInstance.registerUser(SecurityTestEntities.simpleRegistrationRequest)
+
+            assertThat(subscriber.values()).hasSize(1)
+            assertThat(subscriber.values()).first().satisfies {
```
Comment (jhoanmanuelms) [inline submission/submission-security/src/test/kotlin/ebi/ac/uk/security/service/SecurityServiceTest.kt]: Didn't know this one. Nice

File: submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/common/BaseIntegrationTest.kt [hunk 8]
```diff
@@ -11,11 +10,6 @@ internal open class BaseIntegrationTest(private val tempFolder: TemporaryFolder)
     @BeforeAll
     fun init() {
         basePath = tempFolder.root.absolutePath
-        setAppProperty(BASE_PATH_PLACEHOLDER, basePath)
-    }
-
-    @AfterAll
-    fun tearDown() {
-        setAppProperty(basePath, BASE_PATH_PLACEHOLDER)
+        System.setProperty("app.basepath", basePath)
```
Comment (jhoanmanuelms) [inline submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/common/BaseIntegrationTest.kt]: :+1:

File: submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/entities/GenericUser.kt [hunk 9]
```diff
@@ -4,7 +4,7 @@ package ac.uk.ebi.biostd.itest.entities
  * Represent a Generic bio studies user.
  */
 object GenericUser {
-    const val username = "jhon_doe"
+    const val username = "jhon doe"
```
Comment (jhoanmanuelms) [inline submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/entities/GenericUser.kt]: Do we allow spaces in the user name? :thinking:

File: submission/submission-security/src/main/kotlin/ebi/ac/uk/security/integration/SecurityEvents.kt (hunk 3 above)
Comment (Juan-EBI) [reply submission/submission-security/src/main/kotlin/ebi/ac/uk/security/integration/SecurityEvents.kt]: Yes before activation, any suggestion? initially, I called them `activation` and `registration` but that was even easier to confuse for me.

File: submission/submission-security/src/main/kotlin/ebi/ac/uk/security/service/SecurityService.kt (hunk 5 above)
Comment (Juan-EBI) [reply submission/submission-security/src/main/kotlin/ebi/ac/uk/security/service/SecurityService.kt]: mega pro comment

File: submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/entities/GenericUser.kt (hunk 9 above)
Comment (Juan-EBI) [reply submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/entities/GenericUser.kt]: username is actually user name, we need to change this convention in backend and frontend which was inherited from the previous system

**Pull request merged**